File: critical-point/core/src/template/action/idle.rs
use crate::template::action::base::TmplAnimation;
use crate::template::base::impl_tmpl;
use crate::template::variable::TmplVar;
use crate::utils::TmplID;

#[derive(Debug, serde::Serialize, serde::Deserialize, rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
#[rkyv(derive(Debug))]
pub struct TmplActionIdle {
    pub id: TmplID,
    pub enabled: TmplVar<bool>,
    pub character: TmplID,
    pub styles: Vec<TmplID>,
    pub anim_idle: TmplAnimation,
    #[serde(default)]
    pub anim_ready: Option<TmplAnimation>,
    #[serde(default)]
    pub anim_randoms: Vec<TmplAnimation>,
    pub auto_idle_delay: f32,
    pub enter_level: u16,
    pub derive_level: u16,
    pub poise_level: u16,
}

impl_tmpl!(TmplActionIdle, ActionIdle, "ActionIdle");

#[cfg(test)]
mod tests {
    use super::*;
    use crate::template::database::TmplDatabase;
    use crate::utils::{id, LEVEL_IDLE};

    #[test]
    fn test_load_action_idle() {
        let db = TmplDatabase::new(10240, 150).unwrap();

        let act = db.find_as::<TmplActionIdle>(id!("Action.One.Idle")).unwrap();
        assert_eq!(act.id, id!("Action.One.Idle"));
        assert_eq!(act.enabled.value().unwrap(), true);
        assert_eq!(act.character, id!("Character.One"));
        assert_eq!(act.styles.as_slice(), &[id!("Style.One/1"), id!("Style.One/2")]);
        assert_eq!(act.anim_idle.files, "girl_stand_idle.*");
        assert_eq!(act.anim_idle.duration, 2.5);
        assert_eq!(act.anim_idle.fade_in, 0.1);
        assert_eq!(act.anim_idle.root_motion, false);
        let anim_ready = act.anim_ready.as_ref().unwrap();
        assert_eq!(anim_ready.files, "girl_stand_ready.*");
        assert_eq!(anim_ready.duration, 2.0);
        assert_eq!(anim_ready.fade_in, 0.1);
        assert_eq!(anim_ready.root_motion, false);
        assert!(act.anim_randoms.is_empty());
        assert_eq!(act.auto_idle_delay, 10.0);
        assert_eq!(act.enter_level, LEVEL_IDLE);
        assert_eq!(act.derive_level, LEVEL_IDLE);
        assert_eq!(act.poise_level, 0);

        let act2 = db.find_as::<TmplActionIdle>(id!("Action.One.IdleX")).unwrap();
        assert_eq!(act2.id, id!("Action.One.IdleX"));
        assert_eq!(act2.anim_idle.files, "girl_stand_idle.*");
        assert!(act2.anim_ready.is_none());
    }
}

File: turning-point/src/action/idle.ts
import { float, ID, int, parseBool, parseTime } from '../common';
import { Animation, AnimationArgs } from './animation';
import { Action, ActionArgs, LEVEL_IDLE, NpcAction, NpcActionArgs } from './base';

export type ActionIdleArgs = ActionArgs & {
    /** 非战斗状态站立动画 */
    anim_idle: AnimationArgs;

    /** 战斗状态站立动画 */
    anim_ready?: AnimationArgs;

    /** 非战斗状态 随机小动作动画 */
    anim_randoms?: ReadonlyArray<AnimationArgs>;

    /** 从Ready进入Idle的延迟时间 */
    auto_idle_delay?: float | string;

    /** 是否继承上个动作派生 */
    derive_keeping?: boolean | int;
};

export class ActionIdle extends Action {
    /** 非战斗状态站立动画 */
    public readonly anim_idle: Animation;

    /** 战斗状态站立动画 */
    public readonly anim_ready?: Animation;

    /** 非战斗状态 随机小动作动画 */
    public readonly anim_randoms?: ReadonlyArray<Animation>;

    /** 从Ready进入Idle的延迟时间 */
    public readonly auto_idle_delay: float;

    /** 进入等级 */
    public readonly enter_level: int;

    /** 派生等级 */
    public readonly derive_level: int;

    /** 是否继承上个动作派生 */
    public readonly derive_keeping: boolean;

    /** 韧性等级 */
    public readonly poise_level: int;

    public constructor(id: ID, args: ActionIdleArgs) {
        super(id, args);
        this.anim_idle = new Animation(args.anim_idle, this.w('anim_idle'), { root_motion: false });
        this.anim_ready = !args.anim_ready
            ? undefined
            : new Animation(args.anim_ready, this.w('anim_ready'), {
                  root_motion: false,
              });
        this.anim_randoms = !args.anim_randoms
            ? undefined
            : args.anim_randoms.map(
                  (args) =>
                      new Animation(args, this.w('anim_randoms'), {
                          root_motion: false,
                      }),
              );
        this.auto_idle_delay = parseTime(args.auto_idle_delay || '10s', this.w('auto_idle_delay'));
        this.enter_level = LEVEL_IDLE;
        this.derive_level = LEVEL_IDLE;
        this.derive_keeping =
            args.derive_keeping == null
                ? true
                : parseBool(args.derive_keeping, this.w('derive_keeping'));
        this.poise_level = 0;

        Animation.generateLocalID([this.anim_idle, this.anim_ready, ...(this.anim_randoms || [])]);
    }
}

export type NpcActionIdleArgs = NpcActionArgs & {
    /** 非战斗状态站立动画 */
    anim_idle: AnimationArgs;

    /** 战斗状态站立动画 */
    anim_ready?: AnimationArgs;

    /** 从Ready进入Idle的延迟时间 */
    auto_idle_delay?: float | string;
};

export class NpcActionIdle extends NpcAction {
    /** 非战斗状态站立动画 */
    public readonly anim_idle: Animation;

    /** 战斗状态站立动画 */
    public readonly anim_ready?: Animation;

    /** 从Ready进入Idle的延迟时间 */
    public readonly auto_idle_delay: float;

    /** 韧性等级 */
    public readonly poise_level: int;

    public constructor(id: ID, args: NpcActionIdleArgs) {
        super(id, args);
        this.anim_idle = new Animation(args.anim_idle, this.w('anim_idle'), { root_motion: false });
        this.anim_ready = !args.anim_ready
            ? undefined
            : new Animation(args.anim_ready, this.w('anim_ready'), {
                  root_motion: false,
              });
        this.auto_idle_delay = parseTime(args.auto_idle_delay || '10s', this.w('auto_idle_delay'));
        this.poise_level = 0;

        Animation.generateLocalID([this.anim_idle, this.anim_ready]);
    }
}
